File: week1_baseline/ruby/12_context/lib/boukensha/backends/gemini.rb
require_relative "base"

module Boukensha
  module Backends
    class Gemini < Base
      BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models"
      MODELS = {
        "gemini-3.5-flash" => {
          context_window: 1_048_576,
          cost_per_million: { input: 1.5, output: 9.0 },
          usage_unit: :tokens
        },
        "gemini-3.1-flash-lite" => {
          context_window: 1_048_576,
          cost_per_million: { input: 0.25, output: 1.5 },
          usage_unit: :tokens
        },
        # It has
        #"gemini-3.1-pro-preview-customtools" => {
        #  context_window: 1_048_576,
        #  cost_per_million: { input: 2.0, output: 12.0 },
        #  usage_unit: :tokens
        #}
      }.freeze

      def initialize(api_key:, model:)
        @api_key = api_key
        configure_model(model)
      end

      def to_messages(messages)
        messages.map do |msg|
          case msg.role
          when :assistant
            { role: "model", parts: assistant_parts(msg.content) }
          when :tool_result
            {
              role: "user",
              parts: [{
                functionResponse: {
                  name: msg.tool_use_id,
                  response: { content: msg.content }
                }
              }]
            }
          else
            { role: msg.role.to_s, parts: [{ text: msg.content }] }
          end
        end
      end

      def to_tools(tools)
        return [] if tools.empty?

        [{
          functionDeclarations: tools.values.map do |tool|
            {
              name: tool.name,
              description: tool.description,
              parameters: {
                type: "object",
                properties: tool.parameters,
                required: tool.parameters.keys.map(&:to_s)
              }
            }
          end
        }]
      end

      def to_payload(context, max_output_tokens: 1024, tools: nil)
        {
          systemInstruction: { parts: [{ text: context.system }] },
          contents: to_messages(context.messages),
          tools: tools.nil? ? to_tools(context.tools) : tools,
          generationConfig: {
            maxOutputTokens: max_output_tokens,
            thinkingConfig: thinking_config
          }
        }
      end

      def headers
        {
          "Content-Type"   => "application/json",
          "x-goog-api-key" => @api_key
        }
      end

      def url
        "#{BASE_URL}/#{@model}:generateContent"
      end

      # Normalizes a Gemini generateContent response into the common shape:
      #   { stop_reason: "tool_use" | "end_turn", content: [ {"type"=>"text", "text"=>...} | {"type"=>"tool_use", "id"=>, "name"=>, "input"=>} ] }
      #
      # Gemini doesn't assign call ids, so the function name is reused as the
      # id (Gemini also matches functionResponse back to a call by name).
      def parse_response(response)
        parts = response.dig("candidates", 0, "content", "parts") || []

        content   = []
        tool_used = false

        parts.each do |part|
          if part["functionCall"]
            fc = part["functionCall"]
            content << { "type" => "tool_use", "id" => fc["name"], "name" => fc["name"], "input" => fc["args"] || {}, "signature" => part["thoughtSignature"] }
            tool_used = true
          elsif part["thought"]
            content << { "type" => "reasoning", "text" => part["text"].to_s, "signature" => part["thoughtSignature"] }
          elsif part["text"]
            content << { "type" => "text", "text" => part["text"] }
          end
        end

        { stop_reason: tool_used ? "tool_use" : "end_turn", content: content }
      end

      private

      def thinking_config
        case @model
        when "gemini-3.1-pro-preview-customtools"
          { thinkingLevel: "LOW" }   # full disable not supported on this model
        else
          { thinkingBudget: 0 }      # gemini-3.5-flash, gemini-3.1-flash-lite
        end
      end

      # Rebuilds Gemini "model" parts from normalized content blocks
      # (the inverse of parse_response). Text-only turns are stored as a bare
      # String, so wrap it back into a single text block before mapping.
      def assistant_parts(content)
        blocks = content.is_a?(String) ? [{ "type" => "text", "text" => content }] : content

        blocks.map do |b|
          case b["type"]
          when "tool_use"
            part = { functionCall: { name: b["name"], args: b["input"] } }
            part[:thoughtSignature] = b["signature"] if b["signature"]
            part
          when "reasoning"
            part = { text: b["text"].to_s, thought: true }
            part[:thoughtSignature] = b["signature"] if b["signature"]
            part
          else
            { text: b["text"] }
          end
        end
      end
    end
  end
end

File: week1_baseline/python/12_context/boukensha/backends/gemini.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, ClassVar

from .base import Base

if TYPE_CHECKING:
    from ..context import Context
    from ..message import Message
    from ..tool import Tool


class Gemini(Base):
    BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models"
    MODELS: ClassVar[dict[str, dict[str, Any]]] = {
        "gemini-3.5-flash": {
            "context_window": 1_048_576,
            "cost_per_million": {"input": 1.5, "output": 9.0},
            "usage_unit": "tokens",
        },
        "gemini-3.1-flash-lite": {
            "context_window": 1_048_576,
            "cost_per_million": {"input": 0.25, "output": 1.5},
            "usage_unit": "tokens",
        },
        # "gemini-3.1-pro-preview-customtools": {
        #     "context_window": 1_048_576,
        #     "cost_per_million": {"input": 2.0, "output": 12.0},
        #     "usage_unit": "tokens",
        # },
    }

    def __init__(self, *, api_key: str, model: str) -> None:
        self.api_key = api_key
        self._configure_model(model)

    def to_messages(self, messages: list[Message]) -> list[dict[str, Any]]:
        result = []
        for msg in messages:
            if msg.role == "assistant":
                result.append({"role": "model", "parts": self._assistant_parts(msg.content)})
            elif msg.role == "tool_result":
                result.append(
                    {
                        "role": "user",
                        "parts": [
                            {
                                "functionResponse": {
                                    "name": msg.tool_use_id,
                                    "response": {"content": msg.content},
                                }
                            }
                        ],
                    }
                )
            else:
                result.append({"role": msg.role, "parts": [{"text": msg.content}]})
        return result

    def to_tools(self, tools: dict[str, Tool]) -> list[dict[str, Any]]:
        if not tools:
            return []

        return [
            {
                "functionDeclarations": [
                    {
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": {
                            "type": "object",
                            "properties": tool.parameters,
                            "required": list(tool.parameters),
                        },
                    }
                    for tool in tools.values()
                ]
            }
        ]

    def to_payload(
        self, context: Context, *, max_output_tokens: int = 1024, tools: list[dict[str, Any]] | None = None
    ) -> dict[str, Any]:
        return {
            "systemInstruction": {"parts": [{"text": context.system}]},
            "contents": self.to_messages(context.messages),
            "tools": self.to_tools(context.tools) if tools is None else tools,
            "generationConfig": {"maxOutputTokens": max_output_tokens, "thinkingConfig": self._thinking_config()},
        }

    @property
    def headers(self) -> dict[str, str]:
        return {
            "Content-Type": "application/json",
            "x-goog-api-key": self.api_key,
        }

    @property
    def url(self) -> str:
        return f"{self.BASE_URL}/{self.model}:generateContent"

    def parse_response(self, response: dict[str, Any]) -> dict[str, Any]:
        candidates = response.get("candidates") or [{}]
        parts = (candidates[0].get("content") or {}).get("parts") or []

        content: list[dict[str, Any]] = []
        tool_used = False

        for part in parts:
            if part.get("functionCall"):
                fc = part["functionCall"]
                content.append(
                    {
                        "type": "tool_use",
                        "id": fc.get("name"),
                        "name": fc.get("name"),
                        "input": fc.get("args") or {},
                        "signature": part.get("thoughtSignature"),
                    }
                )
                tool_used = True
            elif part.get("thought"):
                content.append(
                    {"type": "reasoning", "text": str(part.get("text") or ""), "signature": part.get("thoughtSignature")}
                )
            elif part.get("text"):
                content.append({"type": "text", "text": part["text"]})

        return {"stop_reason": "tool_use" if tool_used else "end_turn", "content": content}

    def _thinking_config(self) -> dict[str, Any]:
        if self.model == "gemini-3.1-pro-preview-customtools":
            return {"thinkingLevel": "LOW"}  # full disable not supported on this model
        return {"thinkingBudget": 0}  # gemini-3.5-flash, gemini-3.1-flash-lite

    # Rebuilds Gemini "model" parts from normalized content blocks
    # (the inverse of parse_response). Text-only turns are stored as a bare
    # String, so wrap it back into a single text block before mapping.
    @staticmethod
    def _assistant_parts(content: str | list[dict[str, Any]]) -> list[dict[str, Any]]:
        blocks = [{"type": "text", "text": content}] if isinstance(content, str) else content

        parts: list[dict[str, Any]] = []
        for b in blocks:
            if b.get("type") == "tool_use":
                part: dict[str, Any] = {"functionCall": {"name": b["name"], "args": b["input"]}}
                if b.get("signature"):
                    part["thoughtSignature"] = b["signature"]
                parts.append(part)
            elif b.get("type") == "reasoning":
                part = {"text": str(b.get("text") or ""), "thought": True}
                if b.get("signature"):
                    part["thoughtSignature"] = b["signature"]
                parts.append(part)
            else:
                parts.append({"text": b["text"]})
        return parts
